### PR TITLE
Arithmetic expression functions

### PR DESCRIPTION
**Pull request by olavsorl**
## Description

Added +-*/ and changed parsing of JsonTokenType.Number to Decimal instead of Double.
## Related Issue(s)
- [#1179](https://github.com/Altinn/app-frontend-react/issues/1179)

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [x] User documentation is updated with a separate linked PR in [altinn-studio-docs - PR](https://github.com/Altinn/altinn-studio-docs/pull/2667) (if applicable)

## Summary by CodeRabbit

* **New Features**
  * Added arithmetic functions: plus, minus, multiply, divide in expression evaluation
  * Improved numeric parsing to accept scientific notation (e.g., 5E2, 1e-3)

* **Bug Fixes**
  * Better handling and clearer errors for division-by-zero and numeric overflow
  * More consistent numeric representation across expression values

* **Tests**
  * Added comprehensive test suites and shared test infrastructure covering new arithmetic ops and many numeric edge cases

### CODE REVIEW

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/multiply/multiply-multiple-decimals.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Should multiply multiple arguments",
+  "expression": ["multiply", 0.1, 0.2, 15, 1.50],
+  "expects": "0.4500"
```
[the remark this answers is not in the document]
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/multiply/multiply-multiple-decimals.json]: No the test is failing when I set expected to "0.45"

[the remark this answers is not in the document]
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/multiply/multiply-multiple-decimals.json]: Is there a simple way to remove the trailing zeroes?

**Review by TomasEng — COMMENTED** [review]
Er det slik å forstå at både `plus`, `minus`, `multiply` og `divide` kan ta imot et ubegrenset antall operander? Det er ikke noe problem i addisjon og multiplikasjon siden de er kommutative, men for subtraksjon og divisjon synes jeg ikke det gir mening med mer enn to operander (én minuend og én subtrahend for subtraksjon, og én dividend og én divisor for divisjon).  Er det ytret noe ønske om at de skal ta imot flere?

Er det for øvrig gjort noen vurdering av navngivning på de nye funksjonene? For eksempel er jo både `add` og `sum` gode alternativer for `plus`, mens `subtract` og `difference` godt kan brukes i stedet for `minus`. Jeg har ikke noen umiddelbar formening om hva som er best, men tenker det er lurt å vurdere dette grundig siden det blir vanskelig å endre det senere.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide-multiple-decimals.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Should throw exception when more than two arguments is present",
+  "expression": ["divide", 0.1, 0.2, 15, 1.50],
+  "expectsFailure": "Invalid number of args"
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide-multiple-decimals.json]: Denne vil feile i frontend fordi valideringen er satt opp litt annerledes der, og feilmeldingen som kommer er "Invalid expression".  Nå gjør det vel heller ikke noe at feilmeldingene er forskjellige, så lenge det blir kastet en feil både i frontend og i backend. Her er et forslag som går grønt begge steder:
```suggestion
  "expectsFailure": "Invalid"
```
Dette bør være tilstrekkelig for å teste det essensielle, som er at det blir kastet en feil. Det gjelder også de andre, nye operasjonene.

Alternativt må vi sørge for at feilmeldingene er like både i frontend og i backend, men da gjelder dette også andre uttrykk, så jeg tenker det er utenfor omfanget til oppgaven vi skal løse nå.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide-multiple-decimals.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Should throw exception when more than two arguments is present",
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide-multiple-decimals.json]: Liten skrivefeil:
```suggestion
  "name": "Should throw exception when more than two arguments are present",
```
Gjelder også tilsvarende test for de andre operasjonene.

File: src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs [hunk 4]
```diff
@@ -127,6 +127,11 @@ internal static async Task<ExpressionValue> EvaluateExpression_internal(
             ExpressionFunction.argv => Argv(args, positionalArguments),
             ExpressionFunction.gatewayAction => state.GetGatewayAction(),
             ExpressionFunction.language => state.GetLanguage(),
+            // Calculations:
```
Comment (TomasEng) [inline src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs]: Er det meningen at denne kommentaren skal være med?

**Review by TomasEng — CHANGES_REQUESTED** [review]
Jeg har ikke så mye å utsette på C#-koden, men der har kanskje @ivarne mer å bidra med?

Ellers har jeg sett nærmere på de delte testene. Utenom det jeg har kommentert, savner jeg at de dekker litt flere formater for inndata. Jeg ser at de inkluderer både desimaltall og negative tall, og det er veldig bra. Men det mangler hele tall, eksponentformat og strengformat. Hvis jeg skulle skrevet disse testene, ville jeg startet med å sette opp en tabell med forskjellig inndata og forventet resultat. Eksempel for `minus`:
| Minuend | Subtrahend | Ventet resultat |
|:---:|:---:|:---:|
| 3 | 2 | 1 |
| 3.2 | 2 | 1.2 |
| -3 | 2 | -5 |
| 3E2 | 2 | 198 |
| 3E-2 | 2 | -1.97 |
| "3" | 2 | 1 |
| 0 | 2 | -2 |
| null | 2 | null |
| 3 | 2.2 | 0.8 |
| 3 | -2 | 5 |
| 3 | 2E2 | -197 |
| 3 | 2E-2 | 2.98 |
| 3 | "2" | 1 |
| 3 | 0 | 3 |
| 3 | null | null |

Ideen her er å variere én og én variabel og verifisere resultatet. Jeg liker denne måten å strukturere testdataen på fordi det gir god oversikt over hvilke caser som er dekket av testene. Testene for antall argumenter kommer da i tillegg til dette. Vi trenger ikke nødvendigvis gå så grundig til verks, i hvert fall ikke på de kommutative operasjonene, men se gjerne på testene på eksempelvis `greaterThan` for inspirasjon.

File: src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs [hunk 5]
```diff
@@ -589,6 +587,23 @@ private static bool IsSupportedNumericType(Type type)
...
+    {
+        if (
+            double.IsNaN(value)
+            || double.IsInfinity(value)
+            || value > (double)decimal.MaxValue
+            || value < (double)decimal.MinValue
+        )
+        {
+            throw new ExpressionEvaluatorTypeErrorException(
+                $"Cannot convert non-finite or out-of-range number to decimal: {value}"
```
Comment (ivarne) [inline src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs]: Jeg tror dette gjør at vi bør holde oss til `double` som backing data type for expressions, og heller gjøre denne konverteringen i de matematiske uttrykkene. Det som skjer nå er at man ikke kan bruke en double større enn `decimal.MaxValue` til noen ting.

**Comment by olavsorl** [comment]
> Jeg har ikke så mye å utsette på C#-koden, men der har kanskje @ivarne mer å bidra med?
> 
> Ellers har jeg sett nærmere på de delte testene. Utenom det jeg har kommentert, savner jeg at de dekker litt flere formater for inndata. Jeg ser at de inkluderer både desimaltall og negative tall, og det er veldig bra. Men det mangler hele tall, eksponentformat og strengformat. Hvis jeg skulle skrevet disse testene, ville jeg startet med å sette opp en tabell med forskjellig inndata og forventet resultat. Eksempel for `minus`:
> 
> Minuend	Subtrahend	Ventet resultat
> 3	2	1
> 3.2	2	1.2
> -3	2	-5
> 3E2	2	198
> 3E-2	2	-1.97
> "3"	2	1
> 0	2	-2
> null	2	null
> 3	2.2	0.8
> 3	-2	5
> 3	2E2	-197
> 3	2E-2	2.98
> 3	"2"	1
> 3	0	3
> 3	null	null
> Ideen her er å variere én og én variabel og verifisere resultatet. Jeg liker denne måten å strukturere testdataen på fordi det gir god oversikt over hvilke caser som er dekket av testene. Testene for antall argumenter kommer da i tillegg til dette. Vi trenger ikke nødvendigvis gå så grundig til verks, i hvert fall ikke på de kommutative operasjonene, men se gjerne på testene på eksempelvis `greaterThan` for inspirasjon.

Godt forslag! 👍 Jeg har fått Claude til å generere noen flere .json filer. Også har jeg lagt til noen selv og justert litt på navngivning osv. Føler at jeg kan sitte og plukke på disse i en evighet, men det er kanskje ikke noe poeng å bruke så fryktelig mye tid på det hvis det er tilstrekkelig det som er der nå. Lyst til å hjelpe meg å se over @TomasEng?

File: src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs (hunk 4 above)
Comment (olavsorl) [reply src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs]: Kan i grunn fjerne den, har ikke så stor verdi 😄

**Comment by TomasEng** [comment]
Supert, @olavsorl! Jeg skal ta en nærmere kikk, men sånn umiddelbart ser jeg at Coderabbit allerede har lagt inn noen gode kommentarer. I testen som tester `["divide", 10, 3]` tenker jeg det er best å rett og slett velge andre data å teste med, siden vi uavhengig av programmeringsspråk ikke kan gi noe presist resultat på denne (siden vi bruker JSON-format, og JSON har ingen måte å representere brøker på utenom desimaltall). I tilfeller som dette tenker jeg det er akseptabelt at resultatet er forskjellig i frontend og backend, så lenge det er innenfor en rimelig grad av presisjon.

Det er i grunn litt synd at hver testcase må være en egen fil, for det gjør jo terskelen for å legge til nye tester høyere enn den burde være.

File: src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs (hunk 5 above)
Comment (olavsorl) [reply src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs]: Godt poeng! Det kan antageligvis føre til problemer for folk vil jeg tro 👍

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/TestFunctions.cs [hunk 6]
```diff
@@ -357,17 +373,20 @@
...
-            context,
-            positionalArguments
-        );
+        if (test.Expression != null)
+        {
+            await RunTestCaseItem(
+                new ExpressionTestCaseRoot.TestCaseItem()
+                {
+                    Expects = test.Expects,
+                    Expression = test.Expression ?? new Expression(),
```
[the remark this answers is not in the document]
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/TestFunctions.cs]: The compiler does not like this change

File: test/Altinn.App.Core.Tests/TestUtils/TestCasesAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        var theoryData = new List<ExpressionTestCaseRoot.TestCaseItem>();
+        foreach (var file in files)
+        {
+            var data = File.ReadAllText(file);
+            var rootCases = JsonSerializer.Deserialize<ExpressionTestCaseRoot>(data, _jsonSerializerOptions)!;
+            if (rootCases.TestCases is not null)
+            {
+                theoryData.AddRange(rootCases.TestCases);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/TestUtils/TestCasesAttribute.cs:34]: I think readability is better as is.

File: src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs [hunk 8]
```diff
@@ -960,4 +1013,47 @@ private static ExpressionValue Argv(ExpressionValue[] args, ExpressionValue[]? p
...
+    )
+    {
+        if (aDecimal.HasValue is false || bDecimal.HasValue is false)
+        {
+            return null;
+        }
+
+        var result = operation(aDecimal.Value, bDecimal.Value);
+
+        return (double)result;
```
[the remark this answers is not in the document]
Comment (olavsorl) [reply src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs]: We know that this is an issue, and we plan to document this in Altinn Docs.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    {
+      "name": "Should handle very small numbers",
+      "expression": ["divide", 0.00000001, 0.0001],
+      "expects": 0.0001
+    },
+    {
+      "name": "Should divide string scientific notation by number",
+      "expression": ["divide", "5e2", 10],
+      "expectsFailure": "Expected number, got value \"5e2\""
+    },
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:163]: Denne burde vel ikke feile?

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    {
+      "name": "Should fail with invalid string that cannot be parsed as number",
+      "expression": ["divide", 100, "not a number"],
+      "expectsFailure": "Expected number, got value \"not a number\""
+    },
+    {
+      "name": "Should fail with string with no decimals after dot",
+      "expression": ["divide", "55.", 5],
+      "expectsFailure": "Expected number, got value \"55.\""
+    },
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:88]: Dette formatet er gyldig i Javascript. Jeg ser at det her er en regex som gjør det ugyldig. Vil `TryParse` feile hvis man prøver å konvertere `"55."` til et tall?

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    {
+      "name": "Should throw exception when any argument is float max value",
+      "expression": ["divide", 3.4028235E+38, 0],
+      "expectsFailure": "Cannot convert non-finite or out-of-range number to decimal"
+    },
+    {
+      "name": "Should throw exception when any argument is float min value",
+      "expression": ["divide", 0, -3.402825e+38],
+      "expectsFailure": "Cannot convert non-finite or out-of-range number to decimal"
+    }
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Flere observasjoner her:
- Jeg ser at testene venter samme resultat uavhengig av om tallet er i størrelsesorden $`10^{38}`$ eller $`10^{308}`$. Er det riktig, eller hadde du tenkt at de største tallene skulle være representert som strenger?
- Testene dekker de største mulige enhetene (i størrelsesorden `E+38`/`E+308`), men det er ingen tester for de _minste_ mulige enhetene (altså de med negativ eksponent).
- I to av disse testene brukes `0` som divisor, som også er ugyldig. Jeg foreslår å bruke et annet tall, slik at det bare er én ting som er ugyldig i uttrykket.
- Feilmeldingene inneholder "decimal", som er en C#-spesifikk type. Biblioteket vi bruker i Javascript heter tilfeldigvis også "decimal", men det støtter også tall i denne størrelsesordenen, så å si at det ikke er mulig blir ikke helt riktig i frontend. Vi kan godt legge til rette for den samme begrensningen i frontend slik at det blir likt, men da ville jeg gjort feilmeldingen mer generisk, for eksempel "Number is not within the supported range".

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    {
+      "name": "Should handle very large numbers",
+      "expression": ["divide", 1000000000000, 1000000],
+      "expects": 1000000
+    },
+    {
+      "name": "Should handle very large number formatted as string",
+      "expression": ["divide", "9007199254740992", "9007199254740992"],
+      "expects": 1
+    },
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:98]: Her hadde det vært enda mer interessant å teste noe som ville feilet om tallet var formatert som et tall. Så vidt jeg kan se vil `9007199254740992 / 9007199254740992` gi riktig svar (`1`) både med `number` i Javascript og med `double` i C#. Her kunne vi for eksempel prøvd med `["divide", "9999999999999999", 3]`, som bør gi nøyaktig `3333333333333333`. Hvis vi gjør denne operasjonen med `number` i Javascript eller `double` i C#, får vi et annet svar siden $`9999999999999999`$ blir rundet av til $`10^{16}`$.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/minus/minus.json [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      "expression": ["minus", null, null],
+      "expects": null
+    },
+    {
+      "name": "Should return null when one of the arguments is null",
+      "expression": ["minus", null, 2.1],
+      "expects": null
+    },
+    {
+      "name": "Should subtract two integers",
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/minus/minus.json]:
```suggestion
      "name": "Should subtract one integer from another",
```

**Review by TomasEng — COMMENTED** [review]
Nå var testdekningen mye bedre! Jeg har sett over testene spesielt med fokus på hvordan dette vil fungere i frontend. Jeg har kommentert på noen detaljer i `divide`-testene, men det samme gjelder stort sett for alle operasjonene.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json (hunk 9 above)
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:163]: Håpet jeg kunne gjøre det litt enkelt for meg selv ^^ Skal se på det

Comment (TomasEng) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:163]: Hvis det er vanskelig å få til dette i C#, kan vi godt si at vi ikke støtter dette formatet, men da bør vi sette samme begrensning i frontend.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json (hunk 10 above)
Comment (TomasEng) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:88]: Vi har allerede en tilsvarende test på `greaterThan`, så vi lar denne være.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json (hunk 9 above)
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:163]: Har justert regexen lokalt her nå og lagt til en del tester for den. Kommer snart oppdatering i branchen😊

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json (hunk 11 above)
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: * Det er riktig, tanken var å representere float og double her, men det er kanskje litt voldsomt. Foreslår at vi gjør noe sånt som at vi bare tar decimal max og øke bare så mye man må for å få exception til å kastes og motsatt for decimal min. Decimal min er -7.9228162514264338E+28, i testen kan vi bruke en verdi som ligger rett over dette. Så holde oss til to tester. Vi burde kanskje også ha tester for de høyeste tallene som det er støttet at man kan sende inn?
* Ja, tanken med disse testene var å teste at det kastes exception når man er utenfor max og min. Med negativ eksponent så ender man opp med desimaler, eks 1.2E-10 blir vell 0.0000000012 elns, enig i at det gir mening å ha test for det og.
* Jau, enig 👍 
* Ja, det gir vell mening at man validerer mot de grensene man når først. Siden backend tolererer mindre enn frontend i denne sammenheng er det vell god praksis å sette restriksjonene ut ifra hva som er tillatt i backend. Jeg fikser feilmeldingen 👍

Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Hmm, eller er presisjonen her så dårlig at man aldri egentlig bør sende inn så store tallverdier? Resultatet fra disse testene er veldig feil pga konverteringen som skjer mellom double og decimal. 
 Eksempel:
```
    {
      "name": "Should accept value just above decimal min value",
      "expression": ["divide", -7.92281625142643288E+28, 1],
      "expects": -7.9228162514264302E+28
    }
```

Comment (TomasEng) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Ja, [double har en begrensning på 15-17 gjeldende sifre](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types#characteristics-of-the-floating-point-types), så dette tallet vil bli avrundet idet det blir serialisert. Vi kan ikke regne med å få helt nøyaktige svar når vi behandler tall som dette, men jeg tror ikke vi trenger å hindre brukeren fra å gjøre det, for det blir jo bare avrundet. Dessuten vil vi i praksis ikke få inndata som dette, ikke bare fordi det utenfor hva vi bør forvente i vanlige brukstilfeller, men også fordi den samme avrundingen skjer idet dataen blir produsert.

> Ja, tanken med disse testene var å teste at det kastes exception når man er utenfor max og min.

Jepp, og overført til logaritmisk skala er det snakk om max og min på akkurat samme måte. Både grunntallet og eksponenten har en gitt gyldighetsskala.

Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Hadde man hatt en grense for antall desimaler som er tillatt å sende inn, så hadde det vært enklere å sette en grense her. Eller kunne man kanskje validert på antall signifikante tall og satt grensen til 15.  
Resultatet her blir fort litt feil uansett avhengig av hvilke matematisk operasjon det er som utføres. Multipliserer man sammen to tall som er opp mot grensen av antall signifikante tillatt kan man fort få et ganske stort avvik vil jeg tro.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    {
+      "name": "Should handle very large numbers",
+      "expression": ["divide", 1000000000000, 1000000],
+      "expects": 1000000
+    },
+    {
+      "name": "Should handle very large number formatted as string",
+      "expression": ["divide", "9999999999999999", 3],
+      "expects": 3333333333333333
+    },
```
[the remark this answers is not in the document]
Comment (TomasEng) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json:98]: Yes, but we are converting strings directly to `decimal` in C# and we are going to use the `decimal.js` library in Javascript. Verifying that is kind of the point of this test.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json (hunk 11 above)
Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Jeg så ikke at du hadde svart på denne før jeg skrev kommentaren over her ^^ Vet ikke om jeg helt forstår hva det er du mener med avrunding, er det noe decimal.js gjør som ikke skjer i backend, eller er det noe jeg har misforstått?
Det er veldig mange forskjellige typer data som lastes inn i appene. Vet ikke om jeg er så sikker på at det ikke er mulig at vi får inn data som dette. 15 signifikante tall vil jo si at tallene kan bestå av hvilke som helst kombinasjon av heltall og desimaltall.

Comment (TomasEng) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Det har ikke så mye for seg å se på antall tall på høyre og venstre side av kommaet, siden tallene behandles på en logaritmisk skala (altså med et grunntall og en eksponent). Det som gjør det hele litt kontraintuitivt er at tallene er binære, noe som gjør at $`9007199254740993`$ blir rundet av til $`9007199254740992`$, mens $`9007199254740994`$ er gyldig. Tilsvarende hvis vi plasserer kommaet på en vilkårlig plass: $`90071992,54740993`$ blir rundet av til $`90071992,54740994`$. Dette skjer altså idet tallet blir lagret som en `double` i C# eller `number` i Javascript, så selve operasjonen er ikke relevant.

Begrensningen for antall gjeldende sifre er heller ikke grunnen til at vi vil konvertere til `decimal`. Grunnen er at når man gjør operasjoner med `double` eller `number`, så blir ikke det binære tallet avrundet til noe som kan representeres i titallssystemet, noe som gjør at vi får ukorrekt oppførsel som at `["equals", ["plus", 0.1, 0.2], 0.3]` blir `false`.

Så, når det gjelder gjeldende sifre, tror jeg ikke det er nødvendig å sette en begrensning fra vår side. Tallene blir bare rundet av, og det tenker jeg er greit. Vi har allerede den samme oppførselen alle andre steder hvor vi behandler tall. Det som _kan_ gi mening er å gi en feil der det uansett ville bli kastet en feil, som når eksponenten er for stor eller for liten. Men når det gjelder `double` gjør vi ikke det noen andre steder (som jeg vet om), og feilmeldingen som kommer fra C# er i grunn beskrivende nok (i Javascript blir det før øvrig ikke kastet noen feil overhodet; tallene blir bare til `Infinity` ved for stor eksponent og `0` ved for liten eksponent). Når det gjelder casting til `decimal`, er jeg mer usikker på hvilke begrensninger som gjelder. Hva skjer hvis man prøver å caste en `double` i størrelsesorden $`10^{200}`$ eller $`10^{-200}`$ til `decimal` i C#? Det er muligens ting vi må ta høyde for. Med `decimal.js` i Javascript er det ingen begrensning, så her må C# bli styrende.

Comment (olavsorl) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: Vi har blitt enige om å ikke håndtere disse casene eksplisitt i første omgang

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+      "expression": ["divide", 100, "not a number"],
+      "expectsFailure": "Expected number, got value \"not a number\""
+    },
+    {
+      "name": "Should fail with string with no decimals after dot",
+      "expression": ["divide", "55.", 5],
+      "expectsFailure": "Expected number, got value \"55.\""
+    },
+    {
+      "name": "Should handle very large numbers",
```
Comment (TomasEng) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/shared-tests/functions/divide/divide.json]: "Very large" er relativt. Hva med å spesifisere at dette gjelder innenfor begrensningen av tall som kan spesifiseres nøyaktig? (Samme kommentar på tilsvarende tester.)
```suggestion
      "name": "Should handle very large numbers within the safe range",
```

**Review by TomasEng — APPROVED** [review]
Jeg går god for testene, men se gjerne på kommentaren jeg la til. Fint at du også har lagt til flere tester på sammenligningsoperasjonene, så vi får verifisert at vi støtter de samme tallformatene overalt.

Et lite innspill til slutt: Jeg ser at alle testfilene for de aritmestiske uttrykkene inneholder minst én operasjon som ville feilet om vi ikke brukte `decimal` (`0.1 + 0.2`, `50.8 - 20.3`, `0.1 * 0.2`, `0.00000001 / 0.0001`), men kanskje vi i tillegg burde ha én test i hver fil som spesifiserer dette eksplisitt? Eksempel:
```json
{
  "name": "Returns expected value for inputs that would return different result when treated as 64-bit binary floats",
  "expression": ["plus", 0.1, 0.2],
  "expects": 0.3
}
```
Da blir det veldig tydelig hvorfor vi har tatt i bruk `decimal`, og vi unngår at dekningen blir fjernet som følge av at noen endrer testdata i en test som tester noe annet.

Ellers regner jeg med at @ivarne også kan tenke seg å ta en ny kikk. Til info: Vi startet med å se på en løsning hvor vi konverterer tallet direkte til `decimal` dersom det er en streng, men dette har vi gått bort fra. I løsningen vi har nå behandler vi inndataen som `double` uavhengig av format, og konverterer til `decimal` kun for å utføre operasjonen. Begrunnelsen bak det er at vi uansett mister presisjonen vi får med `decimal` når vi returnerer en `double`. (Hadde vi returnert `decimal`, ville det vært en annen sak, men da ville vi fått problemer med serialisering i frontend). Da er det bedre å si at vi behandler inndataen likt uavhengig av format, slik det allerede er i de andre uttrykkene. Dessuten er jo grunnen til at vi har dratt inn `decimal` å gjøre utregningen riktig, ikke å støtte tall med større presisjon.

**Review by ivarne — APPROVED** [review]
